File: apps/mobile/android/app/src/main/java/org/quitmate/app/RiskAlertWidgetProvider.java
package org.quitmate.app;

import android.appwidget.AppWidgetProvider;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.widget.RemoteViews;

public class RiskAlertWidgetProvider extends AppWidgetProvider {
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.risk_alert_widget_layout);

            SharedPreferences prefs = context.getSharedPreferences("HomeWidgetPreferences", Context.MODE_PRIVATE);

            int riskScore = prefs.getInt("risk_score", 0);
            String riskLabel = prefs.getString("risk_label", "\u4F4E");
            int cravingIntensity = prefs.getInt("craving_intensity", 0);

            views.setTextViewText(R.id.risk_score_text, String.valueOf(riskScore));
            views.setTextViewText(R.id.risk_label_text, "\u26A0\uFE0F \u98CE\u9669\u7B49\u7EA7 " + riskLabel);
            views.setProgressBar(R.id.craving_bar, 100, cravingIntensity, false);

            // Color based on risk level: green <40, amber 40-70, red >=70
            int riskColor;
            if (riskScore >= 70) riskColor = 0xFFF44444;
            else if (riskScore >= 40) riskColor = 0xFFFFB74D;
            else riskColor = 0xFF4CAF50;
            views.setTextColor(R.id.risk_score_text, riskColor);

            appWidgetManager.updateAppWidget(appWidgetId, views);
        }
    }
}

File: apps/mobile/android/app/src/main/kotlin/org/quitmate/app/RiskAlertWidgetProvider.kt
package org.quitmate.app

import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.util.Log
import android.widget.RemoteViews

class RiskAlertWidgetProvider : AppWidgetProvider() {
    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        for (appWidgetId in appWidgetIds) {
            try {
                updateWidget(context, appWidgetManager, appWidgetId)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to update widget $appWidgetId", e)
            }
        }
    }

    companion object {
        private const val TAG = "RiskAlertWidget"
        const val PREFS_NAME = "RiskAlertWidgetPreferences"
        const val KEY_RISK_SCORE = "risk_score"
        const val KEY_CRAVING = "craving_intensity"

        fun updateWidget(context: Context, manager: AppWidgetManager, id: Int) {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val views = RemoteViews(context.packageName, R.layout.risk_alert_widget_layout)

            val riskScore = prefs.getInt(KEY_RISK_SCORE, 0)
            val craving = prefs.getInt(KEY_CRAVING, 0)

            views.setTextViewText(R.id.risk_score_text, riskScore.toString())
            views.setProgressBar(R.id.craving_bar, 100, craving, false)

            val color = when {
                riskScore < 30 -> "#4CAF50"
                riskScore < 60 -> "#FFC107"
                riskScore < 80 -> "#FF9800"
                else -> "#F44336"
            }
            views.setTextColor(R.id.risk_score_text, android.graphics.Color.parseColor(color))

            manager.updateAppWidget(id, views)
        }
    }
}